packages/repository/test/unit/repositories/relation-factory.unit.ts

```diff
@@ -0,0 +1,60 @@
+import {
+  relationRepositoryFactory,
+  juggler,
+  DefaultCrudRepository,
+  Entity,
+  RelationDefinitionBase,
+  RelationType,
+  ModelDefinition,
+  HasManyDefinition,
+} from '../../../src';
+import * as RelationFactoryNamespace from '../../../src/repositories/relation.repository'; // needed to spy on the factory function
+import {sinon} from '@loopback/testlab';
+
+describe('relationRepositoryFactory', () => {
+  const instanceId = 4;
+  let meta: RelationDefinitionBase;
+  let relationTestRepo: DefaultCrudRepository<Entity, number>;
+  let db: juggler.DataSource;
+
+  before(givenDataSource);
+  before(givenTestRepository);
+
+  it('calls DefaultHasManyEntityCrudRepository constructor when given HasMany metadata', () => {
+    meta = {
+      type: RelationType.hasMany,
+      keyTo: 'hasManyId',
+    };
+    const spy = sinon.spy(
+      RelationFactoryNamespace,
+      'DefaultHasManyEntityCrudRepository',
+    );
```

Comment:
This is modifying global state which I find rather evil. You are not restoring `DefaultHasManyEntityCrudRepository` after the tests, therefore other tests we may write later will see mocked version instead of the real implementation.

If you want to test `relationRepositoryFactory` in isolation, then you need to find a way how to inject `DefaultHasManyEntityCrudRepository` constructor to the factory function. That will allow you to spy on the class locally.

```ts
const spy = sinon.spy(DefaultHasManyEntityCrudRepository);
// the spy should be used as follows:
// new spy(4, meta, orderRepo);

// This assumes that RelationFactory accepts a key/value map of relation repository classes
new RelationFactory({hasMany: spy}).createHasMany(4, meta, orderRepo);
// etc.
```

To be honest, I am not sure how much value there is in such unit tests. Personally, I'd look into ways how to test `relationRepositoryFactory` in an integration style, where we let the factory create a real repository.

---

> I'm trying to set up a spy for DefaultHasManyEntityCrudRepository to write the unit tests for hasManyRepositoryFactory function and am having a hard time.

```
      const spy = sinon.spy(DefaultHasManyEntityCrudRepository);
hasManyRepositoryFactory(4, meta, orderRepo);

      sinon.assert.calledWithNew(spy);
      sinon.assert.calledWith(spy, orderRepo, {customerId: 4});
```

> Strangely enough, spy doesn't seem to work when spying a function that's being used in another. 
> The first assertion will pass, but the second will fail. Is there a gap in my knowledge with sinon as to why this shouldn't work?

`sinon.spy` cannot rewrite function references stored in variables for you, that's a limitation of JavaScript language. Here is an equivalent of what you are trying to do:

```ts
let a = 1;
spy(a); // spy is supposed to set "a" to 2
expect(a).to.equal(2);
```

Reply:
@bajtos I'm not sure if I agree with your proposal to test this function in integration style anymore. If we were to test the factory function integration style, we'd just end up with what we have for the acceptance test right now, where we assert for products of function calls (create/find) of the constrained repository produced by the factory function. I feel this would be better covered by tests on `DefaultHasManyEntityCrudRepository` constructor itself.

For the sake of increasing our coverage, I thought testing whether the factory function passes the correct parameters into the relation repository constructor would somewhat fulfill the factory function's requirement to have tests.

What do you think?